### ISSUE TITLE
Dunst notifications displayed on lockscreen (betterlockscreen, xtrlock)

### ISSUE THREAD

**Issue by vtnpp** [dunst-project/dunst#697]
### Installation info

- Version: `Dunst - A customizable and lightweight notification-daemon 1.4.1 (2019-07-03)`
- Install type: [`dunst`](https://www.archlinux.org/packages/community/x86_64/dunst/) from official Arch repositories
- Distro and version: `Arch Linux`

**Original issue: https://github.com/pavanjadhaw/betterlockscreen/issues/160**

**Asking here since the issue occurs both with [`betterlockscreen`](https://github.com/pavanjadhaw/betterlockscreen/) and [`xtrlock`](https://www.archlinux.org/packages/community/x86_64/xtrlock/).**

---

I'm using Arch Linux with the following packages:

- [`betterlockscreen-git`<sup>AUR</sup>](https://aur.archlinux.org/packages/betterlockscreen-git/)
- [`dunst`](https://www.archlinux.org/packages/community/x86_64/dunst/)
- [`xidlehook`<sup>AUR</sup>](https://aur.archlinux.org/packages/xidlehook/)

My issue is that `dunst` notifications still appear on lockscreen.

I've run `systemctl enable betterlockscreen@$USER` to enable the user service.

If I understand correctly, since the systemd user service is enabled, betterlockscreen does [the following](https://github.com/zv0n/betterlockscreen/blob/b1d488395a7341d9ee71487008e5793981964b87/betterlockscreen#L39) before locking, and [this](https://github.com/zv0n/betterlockscreen/blob/b1d488395a7341d9ee71487008e5793981964b87/betterlockscreen#L67) after unlocking.

I tried to use a lockscreen script replicating these commands instead of using the regular `betterlockscreen -l -t ""` command as a workaround (since the `dunst` notifications were still displayed), but unfortunately it doesn't make any difference.

- `lockscreen.sh`

```
#!/bin/bash

pkill -u "$USER" -USR1 dunst
betterlockscreen -l -t ""
pkill -u "$USER" -USR2 dunst
```

_Note: I also tried with `killall -SIGUSR1 dunst` and `killall -SIGUSR2 dunst` as suggested on the [Arch Wiki](https://wiki.archlinux.org/index.php/Dunst#Disable_dunst_temporarily)._

and here's how it's triggered by [`xidlehook`<sup>AUR</sup>](https://aur.archlinux.org/packages/xidlehook/):

- `~/.xinitrc`

```
#!/bin/bash

dunst &
xset s on &
xset s 600 &
xidlehook \
--not-when-fullscreen \
--not-when-audio \
--timer 300 '~/scripts/lockscreen.sh' '' &
```

So basically, the screen locks after 5 minutes and turns off after another extra 5 minutes. This works perfectly fine, my only issue is that `dunst` notifications are still displayed on the lockscreen.

It's quite difficult to reproduce.

Thanks in advance.

**Comment by bebehei** [comment]
Well, I'm a maintainer of this and I even have the same problem. :see_no_evil: I'm using i3lock as lockscreen and I'm using compton as a compositing manager.

In our research, we found out, that, it's linked to the compositor of your system, which introduces some weird race conditions or so.... :confused:  Ref: i3/i3lock#204 (we might have a similar issue in here, too).

**Comment by tsipinakis** [comment]
Other than the compositor issue the pause/unpause solution mentioned above is my workaround, and it works perfectly in my case.
For why it doesn't work here, I suspect this is a bug in `betterscreenlock` from a quick read-through of the script it calls i3lock without the `--nofork` so i3lock forks immediately on launch which makes it run the unpause command.

**Comment by vtnpp** [comment]
> In our research, we found out, that, it's linked to the compositor of your system,

Thanks for the info, I'll disable `picom` for now and see if that solves the issue.

> Other than the compositor issue the pause/unpause solution mentioned above is my workaround, and it works perfectly in my case.
For why it doesn't work here, I suspect this is a bug in betterscreenlock from a quick read-through of the script it calls i3lock without the --nofork so i3lock forks immediately on launch which makes it run the unpause command.

As I said, the issue also occurs with `xtrlock` ( which doesn't use `i3lock` if I'm not mistaken), invoked with the following script:

`lockscreen.sh` :

```
#!/bin/bash

pkill -u "$USER" -USR1 dunst       # or killall -SIGUSR1 dunst  
xtrlock -b
pkill -u "$USER" -USR2 dunst       # or killall -SIGUSR2 dunst
```

**Comment by vtnpp** [comment]
So I disabled `picom` and I can confirm that not a single notification appeared on my lockscreen (currently using `betterlockscreen`) in the last ~15 hours. They were all displayed after I unlocked.

**Comment by kamushadenes** [comment]
I can confirm I have the same issue running picom. Disabling it solves the problem as well, but unfortunately I need it. Perhaps we should take this to picom?

**Comment by shervinsahba** [comment]
Are there updates to this? I have the same issue with betterlockscreen: v4.0.3 and picom 8.2-1.

**Comment by vishalmadhvi** [comment]
We wants to disable dunst notification on locked xscreensaver and resume notification when xscreen saver is unlocked.

**Comment by fedang** [comment]
is this issue still relevant? it is unclear to me what to do to resolve

**Comment by fwsmit** [comment]
I think this issue is more related to specific lock screens or compositors. Probably not something that can be fixed in dunst
